R/variance_anhecova.R

```diff
@@ -0,0 +1,75 @@
+#' ANHECOVA Covariance
+#'
+#' @param x (`prediction_cf`) Counter-factual prediction.
+#' @param decompose (`flag`) whether to use decompose method to calculate the variance.
+#' @param randomization (`string`) randomization method.
+#' @param ... Not used.
+#'
+#' @return Named covariance matrix.
+#' @export
+vcovAHECOVA <- function(x, decompose = TRUE, randomization = "simple", ...) {
+  assert_class(x, "prediction_cf")
+  assert_flag(decompose)
+  assert_string(randomization)
+  resi <- attr(x, "residual")
+  est <- as.numeric(x)
+  preds <- attr(x, "predictions")
+  var_preds <- var(preds)
+  y <- attr(x, "response")
+  trt <- attr(x, "treatment")
+  pi_t <- as.numeric(table(trt) / length(trt))
+  trt_lvls <- levels(trt)
+  group_idx <- attr(x, "group_idx")
+
+  idx <- split(seq_len(length(trt)), trt)
+  cov_ymu <- vapply(idx, function(is) stats::cov(y[is], preds[is, ]), FUN.VALUE = rep(0, ncol(preds)))
+
+  if (decompose) {
+    vcov_sr <- (vapply(idx, function(is) stats::var(y[is]), FUN.VALUE = 0) + diag(var_preds) - 2 * diag(cov_ymu)) / pi_t
+  } else {
+    vcov_sr <- vapply(idx, function(is) stats::var(resi[is]), FUN.VALUE = 0) / pi_t
+  }
+
+  v <- diag(vcov_sr) + cov_ymu + t(cov_ymu) - var_preds
+  v <- v - h_get_erb(resi, group_idx, trt, pi_t, randomization)
+  ret <- v / length(resi)
+  dimnames(ret) <- list(trt_lvls, trt_lvls)
+  return(ret)
+}
+
+#' @keywords internal
+h_adjust_pi <- function(pi_t) {
+  diag(pi_t) - pi_t %*% t(pi_t)
+}
+
+#' @keywords internal
+h_get_erb <- function(resi, group_idx, trt, pi_t, randomization) {
+  if (length(group_idx) == 1L) {
+    return(0)
+  }
+  if (randomization %in% c("simple", "pocock-simon")) {
```

Comment:
When condition (U) is satisfied, this last term can be dropped ( cannonical link ）

Reply:
we can simplify this later; for now it is at least working correctly